### PR TITLE
My Area Team functionality

### LINKED ISSUE #177: [My Area] Team creation, invites and mangement
From My Area, you should be able to create a team and invite people to it

- [x] Create new Team (when you have a Ticket)
- [x] Invite Member to your Team as Captain
- [x] View open invites of a Team as Captain
- [x] Cancel invites of a Team as Captain
- [x] Remove members from your team as Captain

- [x] View personal invites
- [x] Accept personal invites
- [x] Decline personal invites
- [x] Remove yourself from a Team

### PR DESCRIPTION
**Pull request by skamoen**
This PR adds Team functionality to My Area. Users with a valid Ticket can create a Team. They become Captain of this team and can invite other members to this team, based on username. 

Captains can remove Users from their team and cancel open invites

Users can accept or decline invites, and remove themselves from a Team

**As this is quite a large PR, it needs thorough testing!**. Suggestions welcome.

Fixes #177

### CODE REVIEW

File: app/elements/lancie-my-area/my-area-teams.html [hunk 1]
```diff
@@ -1,54 +1,104 @@
...
       <div class="card-actions">
-        <paper-icon-button id="new-team" icon="social:group-add" disabled></paper-icon-button>
+        <paper-icon-button id="new-team" icon="social:group-add" on-tap="openCreateTeamDialog"></paper-icon-button>
         <paper-tooltip for="new-team" offset="0" animation-delay="1">Create a new team</paper-tooltip>
+        <paper-icon-button id="teamInviteButton" icon="icons:inbox" on-tap="openPersonalInvitesDialog"></paper-icon-button>
+        <paper-tooltip for="teamInviteButton" offset="0" animation-delay="1">View your open invites</paper-tooltip>
       </div>
     </paper-card>
+    <!-- Dialogs -->
+    <!-- Create Team Dialog -->
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: If you include HTML comments to create subsections in your page, you probably need a new element :wink:

File: app/elements/lancie-my-area/my-area-teams.html [hunk 2]
```diff
@@ -57,13 +107,77 @@
...
+        } else if (r.request.status === 403) {
+          this.createTeamMessage = 'You\'re not allowed to create a team. Do you have a ticket?';
+        } else {
+          this.createTeamMessage = 'Something went wrong';
+        }
+      }
+    },
+    submitAcceptInvite: function() {
+      var token = event.target.parentNode.parentNode.parentNode.parentNode.id;
+      console.log(token);
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: Leftover debugging

File: app/elements/lancie-my-area/my-area-teams.html [hunk 3]
```diff
@@ -57,13 +107,77 @@
...
+          this.createTeamMessage = 'Teamname already in use.';
+        } else if (r.request.status === 403) {
+          this.createTeamMessage = 'You\'re not allowed to create a team. Do you have a ticket?';
+        } else {
+          this.createTeamMessage = 'Something went wrong';
+        }
+      }
+    },
+    submitAcceptInvite: function() {
+      var token = event.target.parentNode.parentNode.parentNode.parentNode.id;
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: This is really really not readible and not maintainable. Why was this needed? Would like to get a cleaner implementation for this.

File: app/elements/lancie-my-area/my-area-teams.html [hunk 4]
```diff
@@ -57,13 +107,77 @@
...
+    refreshInvites: function() {
+      this.$.ajaxGetPersonalInvites.generateRequest();
+      this.refreshTeams();
     },
     storeTeams: function(event, request) {
       this.teams = request.response.length > 0 ? request.response : false;
     },
-    attached: function(){
+    storePersonalInvites: function(event, request) {
+      this.personalInvites = request.response.length > 0 ? request.response : false;
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: I thought `request.response` is an object, not an array?

File: app/elements/lancie-my-area/my-area-teams.html [hunk 5]
```diff
@@ -57,13 +107,77 @@
     properties: {
       teams: {
         type: Array
-      }
+      },
+      user: {
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: Can just be `user: Object`

File: app/elements/lancie-my-area/my-area-teams.html [hunk 6]
```diff
@@ -57,13 +107,77 @@
...
+        type: Object
+      },
+      createTeamError: {
+        type: Boolean,
+        value: false
+      },
+      personalInvites: {
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: can just be `personalInvites: Array`

File: app/elements/lancie-my-area/my-area-teams.html [hunk 7]
```diff
@@ -1,54 +1,104 @@
...
+    
     paper-tooltip {
       --paper-tooltip-background: #1A2B43;
       --paper-tooltip-text-color: #FEE474;
     }
-
+    
+    paper-dialog {
+      min-width: 30%;
+      /*--primary-text-color: #FEE474;*/
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: Leftover code I think?

File: app/elements/lancie-my-area/my-area-teams.html [hunk 8]
```diff
@@ -1,54 +1,104 @@
...
+      --paper-dialog-button-color: #1A2B43;
+    }
+    
+    .member-icon {
+      margin-left: 8%;
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: 8% is quite specific, what was the issue here?

File: app/elements/lancie-my-area/my-area-teams.html [hunk 9]
```diff
@@ -1,54 +1,104 @@
...
+    }
+    
+    .errormessage {
+      padding: 10px 20px 9px;
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: Any reason to leave off a pixel on the right?

File: app/elements/lancie-my-area/my-area-teams.html [hunk 10]
```diff
@@ -57,13 +107,77 @@
...
+      this.$.personalInvitesDialog.open();
+    },
+    submitCreateTeam: function() {
+      this.$.ajaxCreateTeam.body = {
+        teamName: this.teamName,
+        captainUsername: this.user.username
+      };
+      this.$.ajaxCreateTeam.generateRequest();
+    },
+    onCreateTeam: function(e, r) {
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams.html]: This is only called in an `on-response` but also contains error handling. Shouldn't that be in an `on-error` function handler?

File: app/elements/lancie-my-area/my-area-teams/team-item.html [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+<link rel="import" href="../../../bower_components/paper-toast/paper-toast.html">
+<link rel="import" href="../../../bower_components/paper-tooltip/paper-tooltip.html">
+<link rel="import" href="../../lancie-login/lancie-ajax.html">
+<link rel="import" href="team-member-item.html">
+<link rel="import" href="my-area-teams-invite.html">
+<dom-module id="team-item">
+  <template>
+    <style include="iron-flex iron-flex-alignment"></style>
+    <style>
+    :host {
```
Comment (TimvdLippe) [inline app/elements/lancie-my-area/my-area-teams/team-item.html]: The CSS declared here is identical to the previous file.

File: app/elements/lancie-my-area/my-area-teams.html (hunk 10 above)
Comment (skamoen) [inline app/elements/lancie-my-area/my-area-teams.html]: Nope? 

```
<dom-module id="lancie-ajax">
  <template>
    <iron-ajax
      id="ajax"
      body="{{body}}"
      params="{{params}}"
      method="{{method}}"
      url="/api/v1/{{refurl}}"
      handle-as="{{handleAs}}"
      last-response="{{lastResponse}}"
      loading="{{loading}}"
      headers='{{createHeader(token, contenttype)}}'
      debounce-duration="300"
      on-response="handleResponse"
      on-error="handleResponse"
      ></iron-ajax>
  </template>
```

**Comment by TimvdLippe** [comment]
Some small comments from a static analysis. Overall looks good! Will run it soonTM and test if it works and how it looks.

The biggest issue is the amount of duplication, some elements are very similar. It is worth looking in how they can be merged together and streamline the code.

**Comment by skamoen** [comment]
Did some initial cleanup based on feedback. Looking forward to your other suggestions!

**Comment by martijn9612** [comment]
Functionality all works very well, it looks good! Maybe add a check that the user is not adding himself to the team? The message you currently receive is quite unclear and generic. You shouldn't even have to send the request to the server in this situation.

File: app/elements/lancie-my-area/my-area-teams/team-item.html [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+      },
+      isCurrentUser: Boolean,
+      isCaptain: Boolean,
+      memberUsername: String,
+      addMemberMessage: String
+    },
+    refreshInvites: function() {
+      this.$.ajaxGetTeamInvites.generateRequest();
+    },
+    openAddMemberDialog: function() {
```
Comment (martijn9612) [inline app/elements/lancie-my-area/my-area-teams/team-item.html]: Shouldn't the fields be set to the default values when the request is successfully sent?

**Comment by martijn9612** [comment]
Also, the resetting of the fields is done in different parts of the code each time. Sometimes after the request is successfully processed, sometimes upon opening the dialog. I think it is preferrable to do this on successfully processing the request.

**Comment by skamoen** [comment]
It's easier to let the API handle double requests. Self-checking would add an unreasonable amount of code for a border-case, in my opinion. I moved the reset code to a consistent location, namely the opening of the dialog.

**Comment by martijn9612** [comment]
image
Is this really the exception we want to give? This exception is shown when there already is a team with the name you are trying to create.

**Comment by martijn9612** [comment]
image
The body of the open invites list, both for the entire team and personal, are in bold font. It is not needed and looks a bit weird.

**Comment by martijn9612** [comment]
Changes all look good, everything seems to work correctly. Just some small styling issues. When these issues are fixed, this PR can definitely be merged!

**Comment by skamoen** [comment]
When Transferring a ticket, the ticket list should be reloaded to reflect changes. Now a refresh is required before i can see the ticket is "Transferring"

File: app/elements/lancie-dialog/lancie-dialog.html [hunk 13]
```diff
@@ -0,0 +1,95 @@
+<link rel="import" href="../../bower_components/paper-dialog/paper-dialog.html">
+<!--
+A comment describing this element
```
Comment (skamoen) [inline app/elements/lancie-dialog/lancie-dialog.html]: Hello boilerplate?

File: app/elements/lancie-dialog/lancie-dialog.html [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+    :host > ::content > .buttons {
+      border-top: 1px solid #e8e8e8;
+      color: rgba(0, 0, 0, 0.87);
+    }
+    </style>
+    <content></content>
+  </template>
+  <script>
+  (function() {
```
Comment (skamoen) [inline app/elements/lancie-dialog/lancie-dialog.html]: I don't think you have to wrap this in a function

File: app/elements/lancie-my-area/my-area-orders/lancie-ticket-item.html [hunk 15]
```diff
@@ -0,0 +1,109 @@
+<link rel="import" href="../../../bower_components/polymer/polymer.html">
+
+<!--
+A comment describing this element
```
Comment (skamoen) [inline app/elements/lancie-my-area/my-area-orders/lancie-ticket-item.html]: Same here, boilerplate code.

File: app/elements/lancie-my-area/my-area-teams.html [hunk 16]
```diff
@@ -1,69 +1,180 @@
...
     }
-
+    
     paper-tooltip {
       --paper-tooltip-background: #1A2B43;
       --paper-tooltip-text-color: #FEE474;
     }
-
+    
+    paper-dialog {
```
Comment (skamoen) [inline app/elements/lancie-my-area/my-area-teams.html]: Do we still use this with lancie-dialog being a thing?

**Comment by skamoen** [comment]
It looks awesome, it works awesome.
